contracts/defi/provider/CompoundV2Provider.sol

```diff
@@ -0,0 +1,405 @@
+        address cToken = CompoundRegistry(_oracles[1]).getCToken(_token);
+        uint shares = CToken(cToken).balanceOf(address(_wallet));
+        redeem(_wallet, cToken, shares.mul(_fraction).div(10000));
+    }
+
+    /**
+     * @dev Get the amount of investment in a given token.
+     * @param _wallet The target wallet.
+     * @param _token The token address.
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     * @return The value in tokens of the investment (including interests) and the time at which the investment can be removed.
+     */
+    function getInvestment(
+        BaseWallet _wallet, 
+        address _token, 
+        address[] calldata _oracles
+    ) 
+        external 
+        view
+        returns (uint256 _tokenValue, uint256 _periodEnd) 
+    {
+        require(_oracles.length == 2, "CompoundV2: invalid oracles length");
+        address cToken = CompoundRegistry(_oracles[1]).getCToken(_token);
+        uint amount = CToken(cToken).balanceOf(address(_wallet));
+        uint exchangeRateMantissa = CToken(cToken).exchangeRateStored();
+        _tokenValue = amount.mul(exchangeRateMantissa).div(10 ** 18);
+        _periodEnd = 0;
+    }
+
+    /* ********************************** Implementation of Loan ************************************* */
+
+    /**
+     * @dev Opens a collateralized loan.
+     * @param _wallet The target wallet.
+     * @param _collateral The token used as a collateral.
+     * @param _collateralAmount The amount of collateral token provided.
+     * @param _debtToken The token borrowed.
+     * @param _debtAmount The amount of tokens borrowed.
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     * @return (optional) An ID for the loan when the provider enables users to create multiple distinct loans.
+     */
+    function openLoan(
+        BaseWallet _wallet, 
+        address _collateral, 
+        uint256 _collateralAmount, 
+        address _debtToken, 
+        uint256 _debtAmount, 
+        address[] calldata _oracles
+    ) 
+        external 
+        returns (bytes32 _loanId) 
+    {
+        require(_oracles.length == 2, "CompoundV2: invalid oracles length");
+        address[] memory markets = new address[](2);
+        markets[0] = CompoundRegistry(_oracles[1]).getCToken(_collateral);
+        markets[1] = CompoundRegistry(_oracles[1]).getCToken(_debtToken);
+        address comptroller = _oracles[0];
+        _wallet.invoke(comptroller, 0, abi.encodeWithSignature("enterMarkets(address[])", markets));
+        mint(_wallet, markets[0], _collateral, _collateralAmount);
+        borrow(_wallet, markets[1], _debtAmount);
+    }
+
+    /**
+     * @dev Closes a collateralized loan by repaying all debts (plus interest). Note that it does not redeem the collateral.
```

Comment:
Probably worth clarifying that this closes *all* Compound loans, for all markets

Reply:
Agreed and done.